### PR TITLE
Convert rest of the bazel jobs to use bazel scenario, create a test pull-test-infra bazel job

### PR DESCRIPTION
**Pull request by krzyzacy**
https://github.com/kubernetes/test-infra/issues/3628

### CODE REVIEW

**Comment by krzyzacy** [comment]
seems to work
```
senlu@senlu:~/work/src/k8s.io/test-infra$ ./scenarios/kubernetes_bazel.py --test=affected
Run: ('bazel', 'clean', '--expunge')
INFO: Starting clean (this may take a while). Consider using --expunge_async if the clean takes more than several minutes.
Run: ('git', 'diff', '--name-only', '--diff-filter=d', 'f3e6202...72aaf26')
Run: ('bazel', 'query', '--noshow_progress', 'set(jobs/config.json\njobs/pull-test-infra-bazel.sh\nkubetest/main.go\nprow/config.yaml\nscenarios/BUILD\nscenarios/kubernetes_bazel.py\nscenarios/kubernetes_bazel_test.py\n)')
...........
Affected packages: //scenarios:kubernetes_bazel_test.py
//scenarios:kubernetes_bazel.py
//scenarios:BUILD
//prow:config.yaml
//kubetest:main.go
//jobs:pull-test-infra-bazel.sh
//jobs:config.json

Run: ('bazel', 'query', '--keep_going', '--noshow_progress', 'kind(.*_binary, rdeps(//..., set(//scenarios:kubernetes_bazel_test.py\n//scenarios:kubernetes_bazel.py\n//scenarios:BUILD\n//prow:config.yaml\n//kubetest:main.go\n//jobs:pull-test-infra-bazel.sh\n//jobs:config.json\n)))')
Run: ('bazel', 'build', '//prow/cmd/mkpj:mkpj\n//prow/cmd/hook:hook\n//kubetest:kubetest\n//jobs:config_sort\n//experiment:generate_tests\n')
INFO: Analysed 0 targets (1 packages loaded).
INFO: Found 0 targets...
INFO: Elapsed time: 0.978s, Critical Path: 0.01s
INFO: Build completed successfully, 1 total action
Run: ('bazel', 'query', '--keep_going', '--noshow_progress', 'kind(test, rdeps(//..., set(//scenarios:kubernetes_bazel_test.py\n//scenarios:kubernetes_bazel.py\n//scenarios:BUILD\n//prow:config.yaml\n//kubetest:main.go\n//jobs:pull-test-infra-bazel.sh\n//jobs:config.json\n)))')
Triggering tests : //verify:verify-all
//verify:verify_boilerplate
//verify:verify-pylint
//scenarios:kubernetes_bazel_test
//prow/config:go_default_test
//prow/cmd/hook:go_default_test
//kubetest:go_default_test
//jobs:config_test
//jenkins:bootstrap_test

Run: ('bazel', 'test', '--test_output=errors', '//verify:verify-all', '//verify:verify_boilerplate', '//verify:verify-pylint', '//scenarios:kubernetes_bazel_test', '//prow/config:go_default_test', '//prow/cmd/hook:go_default_test', '//kubetest:go_default_test', '//jobs:config_test', '//jenkins:bootstrap_test', '')
INFO: Analysed 8 targets (0 packages loaded).
INFO: Found 8 test targets...
INFO: Elapsed time: 24.935s, Critical Path: 22.09s
INFO: Build completed successfully, 117 total actions
//jenkins:bootstrap_test                                                 PASSED in 4.2s
//jobs:config_test                                                       PASSED in 6.2s
//kubetest:go_default_test                                               PASSED in 0.2s
//prow/cmd/hook:go_default_test                                          PASSED in 0.1s
//prow/config:go_default_test                                            PASSED in 1.2s
//scenarios:kubernetes_bazel_test                                        PASSED in 0.1s
//verify:verify-pylint                                                   PASSED in 21.7s
//verify:verify_boilerplate                                              PASSED in 1.0s

Executed 8 out of 8 tests: 8 tests pass.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
Run: ('/usr/local/google/home/senlu/work/src/k8s.io/test-infra/./scenarios/../images/pull_kubernetes_bazel/coalesce.py',)
Success
```

**Comment by krzyzacy** [comment]
/assign @spxtr @ixdy @apelisse 

It's in a separate job now and I'll verify before make the real pr job use it.

**Comment by krzyzacy** [comment]
@apelisse side question: why we need to build modified packages for a second time?

**Comment by ixdy** [comment]
what do you mean by "a second time"?

**Comment by krzyzacy** [comment]
ah, I was referring to https://github.com/kubernetes/test-infra/blob/master/jobs/pull-test-infra-bazel.sh#L43 and did not notice that's a no-target call :-)

File: scenarios/kubernetes_bazel.py [hunk 1]
```diff
@@ -52,15 +56,58 @@ def get_version():
...
+        for install in args.install:
+            if not os.path.isfile(install):
+                raise ValueError('Invalid install path: %s' % install)
+            check('pip', 'install', '-r', install)
+
     check('bazel', 'clean', '--expunge')
     res = 0
     try:
-        if args.build:
+        if args.build is not None:
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: why is this necessary?

File: scenarios/kubernetes_bazel.py [hunk 2]
```diff
@@ -52,15 +56,58 @@ def get_version():
...
             check('bazel', 'build', *args.build.split(' '))
         if args.release:
             check('bazel', 'build', *args.release.split(' '))
         if args.test:
-            check('bazel', 'test', *args.test.split(' '))
+            if args.test == 'affected':
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: maybe rather than using a special test name here, add another flag?

the value of `args.test` could then instead be used in the bazel query to restrict the targets searched, e.g.
`kind(.*_binary, rdeps(set(%s), set(%s)))' % (args.test, packages)`

File: scenarios/kubernetes_bazel.py [hunk 3]
```diff
@@ -52,15 +56,58 @@ def get_version():
...
+                base = os.getenv('PULL_BASE_SHA', '')
+                pull = os.getenv('PULL_PULL_SHA', '')
+                if not base or not pull:
+                    raise ValueError('PULL_BASE_SHA and PULL_PULL_SHA must be set!')
+
+                diff = check_output(
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: The affected tests/binaries/etc code should be put in a separate function

File: scenarios/kubernetes_bazel.py [hunk 4]
```diff
@@ -52,15 +56,58 @@ def get_version():
...
+            if not os.path.isfile(install):
+                raise ValueError('Invalid install path: %s' % install)
+            check('pip', 'install', '-r', install)
+
     check('bazel', 'clean', '--expunge')
     res = 0
     try:
-        if args.build:
+        if args.build is not None:
             check('bazel', 'build', *args.build.split(' '))
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: build should also probably use the "affected binaries" code.

(possibly removing the `kind(.* binary)` wrapper?)

File: scenarios/kubernetes_bazel.py [hunk 5]
```diff
@@ -82,13 +129,18 @@ def main(args):
     check(test_infra('images/pull_kubernetes_bazel/coalesce.py'))
 
     echo_result(res)
-    exit(res)
+    if res != 0:
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: is this needed?

File: scenarios/kubernetes_bazel.py [hunk 6]
```diff
@@ -52,15 +56,58 @@ def get_version():
 
 def main(args):
     """Trigger a bazel build/test run, and upload results."""
+    # pylint:disable=too-many-branches
+    if args.install:
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: do we still need this? I thought the bazel tests are all hermetic now.

File: jobs/pull-kubernetes-bazel.sh [hunk 7]
```diff
@@ -1,54 +0,0 @@
...
-
-# The bazel cache is a bit confused, possibly due to version skew.
-# To remedy, clean the cache.
-# https://github.com/kubernetes/kubernetes/issues/49844
-bazel clean --expunge
-
-make bazel-build && rc=$? || rc=$?
-
-# Clear test.xml so that we don't pick up old results.
-find -L bazel-testlogs -name 'test.xml' -type f -exec rm '{}' +
```
Comment (ixdy) [inline jobs/pull-kubernetes-bazel.sh]: I think this is missing from the scenario right now, but it's still an issue AFAIK (unless you've discovered otherwise).

File: scenarios/kubernetes_bazel.py [hunk 8]
```diff
@@ -52,15 +56,58 @@ def get_version():
...
+                        check('bazel', 'build', buildables)
+
+                    tests = check_output(
+                        'bazel', 'query',
+                        '--keep_going',
+                        '--noshow_progress',
+                        'kind(test, rdeps(//..., set(%s)))' % packages
+                        )
+
+                    print 'Triggering tests : %s' % tests
```
Comment (apelisse) [inline scenarios/kubernetes_bazel.py]: I think this print statement and the [previous](https://github.com/kubernetes/test-infra/pull/3777/files#diff-c012d3ce1b21c59a5e03d594c01fe8c4R88) are not very helpful. This one even more so than the other, since `bazel test` will end-up showing you the list of tested target anyway.

File: jobs/pull-kubernetes-bazel.sh (hunk 7 above)
Comment (krzyzacy) [inline jobs/pull-kubernetes-bazel.sh]: So somewhere uses a `bazel clean` call and other places clean up this `test.xml` explicitly. Are they solving two different issues or the same one?

File: scenarios/kubernetes_bazel.py (hunk 8 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: ack

File: scenarios/kubernetes_bazel.py (hunk 1 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: I need to handle a bare `bazel build` call without any target here.

File: scenarios/kubernetes_bazel.py (hunk 6 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: is it? https://github.com/kubernetes/test-infra/blob/master/jobs/pull-test-infra-bazel.sh#L24-L29 still exists?

File: scenarios/kubernetes_bazel.py (hunk 2 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: ah, sgtm, didn't think about this

File: scenarios/kubernetes_bazel.py (hunk 1 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: I think the correct thing to do is to build affected, then clean up `text.xml`

File: scenarios/kubernetes_bazel.py (hunk 5 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: `sys.exit(0)` will still raise `SystemExit(0)` which I would like to avoid

**Comment by krzyzacy** [comment]
rewrote bunch of the logic, PTAL

File: jobs/pull-kubernetes-bazel.sh (hunk 7 above)
Comment (ixdy) [inline jobs/pull-kubernetes-bazel.sh]: `bazel clean` will clean up the `test.xml` files, but we probably want to avoid `bazel clean` if possible, since it'll prevent useful caching between runs.

File: jobs/config.json [hunk 9]
```diff
@@ -62,7 +62,9 @@
   },
   "ci-kubernetes-bazel-test": {
     "args": [
-      "--test=--test_tag_filters=-integration --flaky_test_attempts=3 //cmd/... //pkg/... //federation/... //plugin/... //third_party/... //hack/... //hack:verify-all //vendor/k8s.io/..."
+      "--test=//cmd/... //pkg/... //federation/... //plugin/... //third_party/... //hack/... //hack:verify-all //vendor/k8s.io/...",
+      "--test-args=--test_tag_filters=-integration",
```
Comment (ixdy) [inline jobs/config.json]: can you only give a single test-arg at a time? or could you do
`"--test-args=--test_tag_filters=-integration --flaky_test_attempts=3"`?

File: scenarios/kubernetes_bazel.py [hunk 10]
```diff
@@ -50,17 +54,80 @@ def get_version():
...
+    """
+    Run a bazel query against target kind, include targets from args.
+
+    Returns a list of kind objects from bazel query.
+    """
+    return check_output(
+        'bazel', 'query',
+        '--keep_going',
+        '--noshow_progress',
+        'kind(%s, rdeps(//..., set(%s)))' % (kind, ' '.join(packages))
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: (read this comment after the one below)

replace `//...` here with (roughly) `"set(%s)" % ' '.join(selected_packages)` where `selected_packages` is `args.build` or `args.test`.

(you may want to rename `packages` here to `changed_packages` to make this code more clear.)

File: scenarios/kubernetes_bazel.py [hunk 11]
```diff
@@ -50,17 +54,80 @@ def get_version():
     with open('bazel-genfiles/version') as fp:
         return fp.read().strip()
 
+def get_affected(base, pull):
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: nit: "changed" is probably a more correct name than "affected" - affected are the packages returned by the rdeps bazel query.

File: prow/config.yaml [hunk 12]
```diff
@@ -640,6 +642,45 @@ presubmits:
...
+          privileged: true
+        env:
+        - name: GOOGLE_APPLICATION_CREDENTIALS
+          value: /etc/service-account/service-account.json
+        volumeMounts:
+        - name: service
+          mountPath: /etc/service-account
+          readOnly: true
+        - name: cache-ssd
+          mountPath: /root/.cache
```
Comment (ixdy) [inline prow/config.yaml]: also set `TEST_TMPDIR` to this

(we need this for all of the bazel scenario jobs, I think)

File: scenarios/kubernetes_bazel.py [hunk 13]
```diff
@@ -50,17 +54,80 @@ def get_version():
...
+            pull = os.getenv('PULL_PULL_SHA', '')
+            if not base or not pull:
+                raise ValueError('PULL_BASE_SHA and PULL_PULL_SHA must be set!')
+            affected = get_affected(base, pull)
+            build_pkgs = affected
+            test_pkgs = affected
+
         if args.build:
-            check('bazel', 'build', *args.build.split(' '))
+            build_pkgs += args.build.split(' ')
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: ah, I think there was a misunderstanding with my previous comment. I think we want to use `args.build` and `args.test` to filter the `affected` list, rather than adding them to the affected list.

basically, the interface I'm expecting is that I tell the bazel scenario all of the test packages I care about - let's say `//foo/...` and `//bar/...`; I never want to run tests under `//baz/...`.

If I just pass `--test='//foo/... //bar/...'` then I want to run all tests there.

If I do `--test='//foo/... //bar/...' --affected` then it should only run affected tests in `//foo/...` and `//bar/...`, but should not run any tests under `//baz/...`, even if there are changes affecting tests under `//baz/...`.

File: scenarios/kubernetes_bazel.py [hunk 14]
```diff
@@ -82,17 +149,27 @@ def main(args):
...
-        '--build', help='Bazel build')
+        '--affected', action='store_true',
+        help='If build/test affected targets.')
+    parser.add_argument(
+        '--build', help='Bazel build targets, split by one space')
+    # TODO(krzyzacy): Convert to bazel build rules
+    parser.add_argument(
+        '--install', action="append", help='Dependency(s) that need to be installed')
+    parser.add_argument(
+        '--release', help='Bazel build and push build, split by one space')
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: I don't understand this help string.

File: scenarios/kubernetes_bazel.py [hunk 15]
```diff
@@ -82,17 +149,27 @@ def main(args):
...
     parser = argparse.ArgumentParser()
     parser.add_argument(
-        '--build', help='Bazel build')
+        '--affected', action='store_true',
+        help='If build/test affected targets.')
+    parser.add_argument(
+        '--build', help='Bazel build targets, split by one space')
+    # TODO(krzyzacy): Convert to bazel build rules
+    parser.add_argument(
+        '--install', action="append", help='Dependency(s) that need to be installed')
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: Python dependency(s) ...

File: scenarios/kubernetes_bazel.py [hunk 16]
```diff
@@ -82,17 +149,27 @@ def main(args):
...
+    if res != 0:
+        sys.exit(res)
+
 
 def create_parser():
     """Create argparser."""
     parser = argparse.ArgumentParser()
     parser.add_argument(
-        '--build', help='Bazel build')
+        '--affected', action='store_true',
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: should at least one of the jobs (maybe test-infra) be using this flag?

File: jobs/config.json (hunk 9 above)
Comment (krzyzacy) [inline jobs/config.json]: One arg at a time, or else we need to split them by space?

File: scenarios/kubernetes_bazel.py (hunk 13 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: ah now it makes sense :-) I completely went the other way instead

File: scenarios/kubernetes_bazel.py (hunk 16 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: yeah made infra-bazel-scenario using that.

File: prow/config.yaml (hunk 12 above)
Comment (krzyzacy) [inline prow/config.yaml]: done

File: scenarios/kubernetes_bazel.py (hunk 14 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: reworded a bit

File: scenarios/kubernetes_bazel.py (hunk 15 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: done

File: scenarios/kubernetes_bazel.py (hunk 13 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: done

File: scenarios/kubernetes_bazel.py (hunk 10 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: done

File: scenarios/kubernetes_bazel.py (hunk 11 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: done

**Comment by krzyzacy** [comment]
PTAL, Flipped the logic back :-)

File: scenarios/kubernetes_bazel.py [hunk 17]
```diff
@@ -50,17 +54,92 @@ def get_version():
...
+            if buildables:
+                check('bazel', 'build', *buildables)
+
         if args.release:
             check('bazel', 'build', *args.release.split(' '))
-        if args.test:
-            check('bazel', 'test', *args.test.split(' '))
+
+        # clean up previous test.xml
+        clean_file_in_dir(test_infra('bazel-testlogs'), 'test.xml')
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: this should probably be `./bazel-testlogs`, not `test_infra('bazel-testlogs')` - e.g. for k/k we want to clean up the kubernetes test logs, not the test-infra ones.

File: scenarios/kubernetes_bazel.py [hunk 18]
```diff
@@ -50,17 +54,92 @@ def get_version():
...
+
+        if test_pkgs or affected:
+            tests = query('test', test_pkgs, affected)
+            if args.test_args:
+                tests = args.test_args + tests
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: I think you meant to delete this? I imagine it should look like the `if buildables:` conditional above.

Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: This is to append test args in front of test targets? The query above filters the targets.

File: scenarios/kubernetes_bazel.py (hunk 17 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: done, and I figured out it was not the right test-infra checkout :-)

File: scenarios/kubernetes_bazel.py (hunk 18 above)
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: you're right. I confused `args.test` and `args.test_args`. 🙃

File: scenarios/kubernetes_bazel.py [hunk 19]
```diff
@@ -50,17 +54,92 @@ def get_version():
...
+            check('bazel', 'test', *tests)
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: that said, we probably want a conditional here to only run this if there are affected tests.

Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: whoops, right :-) done

File: scenarios/kubernetes_bazel.py [hunk 20]
```diff
@@ -50,17 +54,93 @@ def get_version():
...
+            if buildables:
+                check('bazel', 'build', *buildables)
+
         if args.release:
             check('bazel', 'build', *args.release.split(' '))
-        if args.test:
-            check('bazel', 'test', *args.test.split(' '))
+
+        # clean up previous test.xml
+        clean_file_in_dir('./bazel-testlogs', 'test.xml')
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: also, I think we may need to guarantee that bazel has run before we get here, as otherwise this symlink may not exist.

in the shell script we did this with a bazel build with no arguments. maybe fold it into L130 and run bazel build always, regardless of whether there are buildables or whether --build was set?

Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: calling `bazel build` regardless now.

**Comment by krzyzacy** [comment]
PTAL

**Comment by krzyzacy** [comment]
rebased :-) @ixdy

File: scenarios/kubernetes_bazel.py (hunk 16 above)
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: looks like this got lost in rebase?

**Comment by ixdy** [comment]
lgtm otherwise

File: prow/config.yaml [hunk 21]
```diff
@@ -79,9 +79,7 @@ presubmits:
         - "--repo=k8s.io/$(REPO_NAME)=$(PULL_REFS)"
         - "--service-account=/etc/service-account/service-account.json"
         - "--upload=gs://kubernetes-jenkins/pr-logs"
-        env:
-        - name: TEST_TMPDIR
```
Comment (ixdy) [inline prow/config.yaml]: why did these get rermoved?

Comment (krzyzacy) [inline prow/config.yaml]: :-( that was a disappointing rebase

Comment (krzyzacy) [inline prow/config.yaml]: fixed

File: scenarios/kubernetes_bazel.py (hunk 16 above)
Comment (krzyzacy) [inline scenarios/kubernetes_bazel.py]: fixed

**Comment by krzyzacy** [comment]
awwwwwww I figured I have a different local branch on my mac :-(

**Comment by krzyzacy** [comment]
seems cla bot is sleeping today

**Comment by krzyzacy** [comment]
merging and will monitor and pray

**Comment by ixdy** [comment]
oh, I think we broke batch builds:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/batch/pull-kubernetes-bazel/35192/
```
W0810 23:30:13.853] Traceback (most recent call last):
W0810 23:30:13.853]   File "/workspace/./test-infra/jenkins/../scenarios/kubernetes_bazel.py", line 204, in <module>
W0810 23:30:13.854]     main(parse_args())
W0810 23:30:13.854]   File "/workspace/./test-infra/jenkins/../scenarios/kubernetes_bazel.py", line 118, in main
W0810 23:30:13.854]     raise ValueError('PULL_BASE_SHA and PULL_PULL_SHA must be set!')
W0810 23:30:13.854] ValueError: PULL_BASE_SHA and PULL_PULL_SHA must be set!
```

File: scenarios/kubernetes_bazel.py [hunk 22]
```diff
@@ -50,17 +54,94 @@ def get_version():
...
     check('bazel', 'clean', '--expunge')
     res = 0
     try:
+        affected = None
+        if args.affected:
+            base = os.getenv('PULL_BASE_SHA', '')
+            pull = os.getenv('PULL_PULL_SHA', '')
+            if not base or not pull:
+                raise ValueError('PULL_BASE_SHA and PULL_PULL_SHA must be set!')
+            affected = get_changed(base, pull)
```
Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: maybe this should be `$PULL_BASE_SHA` to `HEAD`?

Comment (ixdy) [inline scenarios/kubernetes_bazel.py]: https://github.com/kubernetes/test-infra/pull/4051
